libvast/vast/flow.hpp

```diff
@@ -0,0 +1,73 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <functional>
+
+#include "vast/address.hpp"
+#include "vast/port.hpp"
+
+namespace vast {
+
+/// A connection 5-tuple, consisting of IP addresses and transport-layer ports
+/// for originator and resopnder. The protocol type is encoded in the ports.
+struct flow {
+  address src_addr;
+  address dst_addr;
+  port src_port;
+  port dst_port;
+};
+
+/// @returns the protocol of a flow tuple.
+/// @param x The flow to extract the protocol from.
+/// @relates flow
+port::port_type protocol(const flow& x);
+
+/// @returns a hash value for a flow tuple.
+/// @param x The flow to compute the hash value from.
+/// @relates flow
+size_t hash(const flow& x);
```

Comment:
Could you elaborate on the need for this function? I wonder how much it adds on top of `std::hash<flow>`. This is where I would see the implementation of this going. At least this is what we do with all other types.

Reply:
The extra function allows the implementation to move to the `.cpp` file instead. Since the implementation relies on `uhash` and `xxhash`, this avoids including `vast/concept/hashable/uhash.hpp` and `vast/concept/hashable/xxhash.hpp` in `flow.hpp` in order to minimize header dependencies and speeding up compilation.